File: Data Structures & Algorithms/linked-list-cycle-detection/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        '''
        Time Complexity: O(n) n = len(linked list)
        Space Complexity: O(1)
        '''
        fast, slow = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if fast == slow:
                return True
        
        return False

File: Data Structures & Algorithms/is-palindrome/submission-4.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        // Time Complexity: O(n) Space Complexity: O(1)
        int l{0};
        int r{int(s.size()) - 1};
        while (l < r) {
            while (!isalnum(s[l]) && l < r) l++;
            while (!isalnum(s[r]) && r > l) r--;
            if (tolower(s[l]) != tolower(s[r])) return false;
            l++;
            r--;
        }
        return true;
    }
};
